Enable race detector when running tests

### What platforms are affected?

linux, macos, windows, others

### What architectures are affected?

amd64, arm64, others

### What SpiceDB version are you using?

v.1.25.0

### Steps to Reproduce

1. Modify the `magefiles/test.go` to enable the `-race` flag for unit tests
2. Run the unit tests
3. 2 race conditions are detected


### Expected Result

The race detector should always be running for unit tests with no race conditions detected.

### Actual Result

2 race conditions are detected

````
==================
WARNING: DATA RACE
Write at 0x00c004d8d9ec by goroutine 65329:
  github.com/authzed/spicedb/internal/graph.(*limitTracker).markAlreadyPublished()
      /Users/braden.groom/src/spicedb/internal/graph/limits.go:73 +0x78
  github.com/authzed/spicedb/internal/graph.(*parallelLimitedIndexedStream[go.shape.*uint8]).completedTaskIndex()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:501 +0x304
  github.com/authzed/spicedb/internal/graph.withInternalParallelizedStreamingIterableInCursor[go.shape.struct { github.com/authzed/spicedb/internal/graph.resourceType *github.com/authzed/spicedb/pkg/proto/core/v1.RelationReference; github.com/authzed/spicedb/internal/graph.resourcesAndSubjects github.com/authzed/spicedb/pkg/genutil/mapz.ReadOnlyMultimap[string,github.com/authzed/spicedb/internal/graph.subjectInfo] },go.shape.*uint8].func1()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:384 +0x280
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).runner()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:102 +0x98
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).spawnIfAvailable.func1()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:74 +0x34

Previous read at 0x00c004d8d9ec by goroutine 65330:
  github.com/authzed/spicedb/internal/graph.(*limitTracker).hasExhaustedLimit()
      /Users/braden.groom/src/spicedb/internal/graph/limits.go:84 +0x98
  github.com/authzed/spicedb/internal/graph.withInternalParallelizedStreamingIterableInCursor[go.shape.struct { github.com/authzed/spicedb/internal/graph.resourceType *github.com/authzed/spicedb/pkg/proto/core/v1.RelationReference; github.com/authzed/spicedb/internal/graph.resourcesAndSubjects github.com/authzed/spicedb/pkg/genutil/mapz.ReadOnlyMultimap[string,github.com/authzed/spicedb/internal/graph.subjectInfo] },go.shape.*uint8].func1()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:361 +0x74
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).runner()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:102 +0x98
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).spawnIfAvailable.func1()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:74 +0x34

Goroutine 65329 (running) created at:
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).spawnIfAvailable()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:74 +0x114
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).Start()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:52 +0x58
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).StartAndWait()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:58 +0x38
  github.com/authzed/spicedb/internal/graph.withInternalParallelizedStreamingIterableInCursor[go.shape.struct { github.com/authzed/spicedb/internal/graph.resourceType *github.com/authzed/spicedb/pkg/proto/core/v1.RelationReference; github.com/authzed/spicedb/internal/graph.resourcesAndSubjects github.com/authzed/spicedb/pkg/genutil/mapz.ReadOnlyMultimap[string,github.com/authzed/spicedb/internal/graph.subjectInfo] },go.shape.*uint8]()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:388 +0x44c
  github.com/authzed/spicedb/internal/graph.withDatastoreCursorInCursor[go.shape.struct { github.com/authzed/spicedb/internal/graph.resourceType *github.com/authzed/spicedb/pkg/proto/core/v1.RelationReference; github.com/authzed/spicedb/internal/graph.resourcesAndSubjects github.com/authzed/spicedb/pkg/genutil/mapz.ReadOnlyMultimap[string,github.com/authzed/spicedb/internal/graph.subjectInfo] },go.shape.*uint8]()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:200 +0x498
  github.com/authzed/spicedb/internal/graph.(*CursoredReachableResources).redispatchOrReportOverDatabaseQuery()
      /Users/braden.groom/src/spicedb/internal/graph/reachableresources.go:265 +0x1d8
  github.com/authzed/spicedb/internal/graph.(*CursoredReachableResources).lookupRelationEntrypoint()
      /Users/braden.groom/src/spicedb/internal/graph/reachableresources.go:225 +0x7c0
  github.com/authzed/spicedb/internal/graph.(*CursoredReachableResources).afterSameType.func1()
      /Users/braden.groom/src/spicedb/internal/graph/reachableresources.go:129 +0x428
  github.com/authzed/spicedb/internal/graph.withInternalParallelizedStreamingIterableInCursor[go.shape.struct { github.com/authzed/spicedb/pkg/typesystem.re *github.com/authzed/spicedb/pkg/proto/core/v1.ReachabilityEntrypoint; github.com/authzed/spicedb/pkg/typesystem.parentRelation *github.com/authzed/spicedb/pkg/proto/core/v1.RelationReference },go.shape.*uint8].func1()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:374 +0x1d0
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).runner()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:102 +0x98
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).spawnIfAvailable.func1()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:74 +0x34

Goroutine 65330 (running) created at:
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).spawnIfAvailable()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:74 +0x114
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).Start()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:52 +0x58
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).StartAndWait()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:58 +0x38
  github.com/authzed/spicedb/internal/graph.withInternalParallelizedStreamingIterableInCursor[go.shape.struct { github.com/authzed/spicedb/internal/graph.resourceType *github.com/authzed/spicedb/pkg/proto/core/v1.RelationReference; github.com/authzed/spicedb/internal/graph.resourcesAndSubjects github.com/authzed/spicedb/pkg/genutil/mapz.ReadOnlyMultimap[string,github.com/authzed/spicedb/internal/graph.subjectInfo] },go.shape.*uint8]()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:388 +0x44c
  github.com/authzed/spicedb/internal/graph.withDatastoreCursorInCursor[go.shape.struct { github.com/authzed/spicedb/internal/graph.resourceType *github.com/authzed/spicedb/pkg/proto/core/v1.RelationReference; github.com/authzed/spicedb/internal/graph.resourcesAndSubjects github.com/authzed/spicedb/pkg/genutil/mapz.ReadOnlyMultimap[string,github.com/authzed/spicedb/internal/graph.subjectInfo] },go.shape.*uint8]()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:200 +0x498
  github.com/authzed/spicedb/internal/graph.(*CursoredReachableResources).redispatchOrReportOverDatabaseQuery()
      /Users/braden.groom/src/spicedb/internal/graph/reachableresources.go:265 +0x1d8
  github.com/authzed/spicedb/internal/graph.(*CursoredReachableResources).lookupRelationEntrypoint()
      /Users/braden.groom/src/spicedb/internal/graph/reachableresources.go:225 +0x7c0
  github.com/authzed/spicedb/internal/graph.(*CursoredReachableResources).afterSameType.func1()
      /Users/braden.groom/src/spicedb/internal/graph/reachableresources.go:129 +0x428
  github.com/authzed/spicedb/internal/graph.withInternalParallelizedStreamingIterableInCursor[go.shape.struct { github.com/authzed/spicedb/pkg/typesystem.re *github.com/authzed/spicedb/pkg/proto/core/v1.ReachabilityEntrypoint; github.com/authzed/spicedb/pkg/typesystem.parentRelation *github.com/authzed/spicedb/pkg/proto/core/v1.RelationReference },go.shape.*uint8].func1()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:374 +0x1d0
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).runner()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:102 +0x98
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).spawnIfAvailable.func1()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:74 +0x34
==================
--- FAIL: TestReachableResourcesPaginationWithLimit (1.18s)
    --- FAIL: TestReachableResourcesPaginationWithLimit/limit-1 (1.11s)
        testing.go:1465: race detected during execution of test
    testing.go:1465: race detected during execution of test
FAIL
FAIL	github.com/authzed/spicedb/internal/dispatch/graph	111.231s
ok  	github.com/authzed/spicedb/internal/dispatch/keys	1.833s
ok  	github.com/authzed/spicedb/internal/dispatch/remote	1.588s
ok  	github.com/authzed/spicedb/internal/gateway	1.625s
==================
WARNING: DATA RACE
Read at 0x00c00001034c by goroutine 238:
  github.com/authzed/spicedb/internal/graph.(*limitTracker).hasExhaustedLimit()
      /Users/braden.groom/src/spicedb/internal/graph/limits.go:84 +0x88
  github.com/authzed/spicedb/internal/graph.withInternalParallelizedStreamingIterableInCursor[go.shape.int,go.shape.int].func1()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:361 +0x64
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).runner()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:102 +0x98
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).spawnIfAvailable.func1()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:74 +0x34

Previous write at 0x00c00001034c by goroutine 237:
  github.com/authzed/spicedb/internal/graph.(*limitTracker).prepareForPublishing()
      /Users/braden.groom/src/spicedb/internal/graph/limits.go:58 +0x51c
  github.com/authzed/spicedb/internal/graph.(*parallelLimitedIndexedStream[go.shape.int]).completedTaskIndex()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:512 +0x484
  github.com/authzed/spicedb/internal/graph.withInternalParallelizedStreamingIterableInCursor[go.shape.int,go.shape.int].func1()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:384 +0x270
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).runner()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:102 +0x98
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).spawnIfAvailable.func1()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:74 +0x34

Goroutine 238 (running) created at:
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).spawnIfAvailable()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:74 +0x114
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).Start()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:52 +0x58
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).StartAndWait()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:58 +0x38
  github.com/authzed/spicedb/internal/graph.withInternalParallelizedStreamingIterableInCursor[go.shape.int,go.shape.int]()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:388 +0x3d4
  github.com/authzed/spicedb/internal/graph.withParallelizedStreamingIterableInCursor[go.shape.int,go.shape.int]()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:329 +0x2b8
  github.com/authzed/spicedb/internal/graph.TestWithParallelizedStreamingIterableInCursorWithLimit()
      /Users/braden.groom/src/spicedb/internal/graph/cursors_test.go:246 +0x2b8
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.21.1/libexec/src/testing/testing.go:1595 +0x194
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.21.1/libexec/src/testing/testing.go:1648 +0x40

Goroutine 237 (finished) created at:
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).spawnIfAvailable()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:74 +0x114
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).Start()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:52 +0x58
  github.com/authzed/spicedb/internal/taskrunner.(*PreloadedTaskRunner).StartAndWait()
      /Users/braden.groom/src/spicedb/internal/taskrunner/preloadedtaskrunner.go:58 +0x38
  github.com/authzed/spicedb/internal/graph.withInternalParallelizedStreamingIterableInCursor[go.shape.int,go.shape.int]()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:388 +0x3d4
  github.com/authzed/spicedb/internal/graph.withParallelizedStreamingIterableInCursor[go.shape.int,go.shape.int]()
      /Users/braden.groom/src/spicedb/internal/graph/cursors.go:329 +0x2b8
  github.com/authzed/spicedb/internal/graph.TestWithParallelizedStreamingIterableInCursorWithLimit()
      /Users/braden.groom/src/spicedb/internal/graph/cursors_test.go:246 +0x2b8
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.21.1/libexec/src/testing/testing.go:1595 +0x194
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.21.1/libexec/src/testing/testing.go:1648 +0x40
==================
--- FAIL: TestWithParallelizedStreamingIterableInCursorWithLimit (0.00s)
    testing.go:1465: race detected during execution of test
FAIL
FAIL	github.com/authzed/spicedb/internal/graph	0.496s
````

--- a/internal/datastore/common/gc.go
+++ b/internal/datastore/common/gc.go
@@ -99,10 +99,10 @@ var MaxGCInterval = 60 * time.Minute
 // StartGarbageCollector loops forever until the context is canceled and
 // performs garbage collection on the provided interval.
 func StartGarbageCollector(ctx context.Context, gc GarbageCollector, interval, window, timeout time.Duration) error {
-	return startGarbageCollectorWithMaxElapsedTime(ctx, gc, interval, window, 0, timeout)
+	return startGarbageCollectorWithMaxElapsedTime(ctx, gc, interval, window, 0, timeout, gcFailureCounter)
 }
 
-func startGarbageCollectorWithMaxElapsedTime(ctx context.Context, gc GarbageCollector, interval, window, maxElapsedTime, timeout time.Duration) error {
+func startGarbageCollectorWithMaxElapsedTime(ctx context.Context, gc GarbageCollector, interval, window, maxElapsedTime, timeout time.Duration, failureCounter prometheus.Counter) error {
 	backoffInterval := backoff.NewExponentialBackOff()
 	backoffInterval.InitialInterval = interval
 	backoffInterval.MaxInterval = max(MaxGCInterval, interval)
@@ -130,7 +130,7 @@ func startGarbageCollectorWithMaxElapsedTime(ctx context.Context, gc GarbageColl
 
 			err := RunGarbageCollection(gc, window, timeout)
 			if err != nil {
-				gcFailureCounter.Inc()
+				failureCounter.Inc()
 				nextInterval = backoffInterval.NextBackOff()
 				log.Ctx(ctx).Warn().Err(err).
 					Dur("next-attempt-in", nextInterval).

--- a/internal/datastore/common/gc_test.go
+++ b/internal/datastore/common/gc_test.go
@@ -32,16 +32,14 @@ func (t testGC) DeleteBeforeTx(_ context.Context, _ datastore.Revision) (Deletio
 }
 
 func TestGCFailureBackoff(t *testing.T) {
-	defer func() {
-		gcFailureCounter = prometheus.NewCounter(gcFailureCounterConfig)
-	}()
+	localCounter := prometheus.NewCounter(gcFailureCounterConfig)
 	reg := prometheus.NewRegistry()
-	require.NoError(t, reg.Register(gcFailureCounter))
+	require.NoError(t, reg.Register(localCounter))
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 	go func() {
-		require.Error(t, startGarbageCollectorWithMaxElapsedTime(ctx, testGC{}, 100*time.Millisecond, 1*time.Second, 1*time.Nanosecond, 1*time.Minute))
+		require.Error(t, startGarbageCollectorWithMaxElapsedTime(ctx, testGC{}, 100*time.Millisecond, 1*time.Second, 1*time.Nanosecond, 1*time.Minute, localCounter))
 	}()
 	time.Sleep(200 * time.Millisecond)
 	cancel()
@@ -56,13 +54,13 @@ func TestGCFailureBackoff(t *testing.T) {
 	}
 	require.Greater(t, *(mf.GetMetric()[0].Counter.Value), 100.0, "MaxElapsedTime=1ns did not cause backoff to get ignored")
 
-	gcFailureCounter = prometheus.NewCounter(gcFailureCounterConfig)
+	localCounter = prometheus.NewCounter(gcFailureCounterConfig)
 	reg = prometheus.NewRegistry()
-	require.NoError(t, reg.Register(gcFailureCounter))
+	require.NoError(t, reg.Register(localCounter))
 	ctx, cancel = context.WithCancel(context.Background())
 	defer cancel()
 	go func() {
-		require.Error(t, StartGarbageCollector(ctx, testGC{}, 100*time.Millisecond, 1*time.Second, 1*time.Minute))
+		require.Error(t, startGarbageCollectorWithMaxElapsedTime(ctx, testGC{}, 100*time.Millisecond, 0, 1*time.Second, 1*time.Minute, localCounter))
 	}()
 	time.Sleep(200 * time.Millisecond)
 	cancel()

--- a/internal/datastore/memdb/memdb_test.go
+++ b/internal/datastore/memdb/memdb_test.go
@@ -94,7 +94,7 @@ func TestConcurrentWriteRelsError(t *testing.T) {
 	for i := 0; i < 50; i++ {
 		i := i
 		g.Go(func() error {
-			_, err = ds.ReadWriteTx(ctx, func(rwt datastore.ReadWriteTransaction) error {
+			_, err := ds.ReadWriteTx(ctx, func(rwt datastore.ReadWriteTransaction) error {
 				updates := []*corev1.RelationTupleUpdate{}
 				for j := 0; j < 500; j++ {
 					updates = append(updates, &corev1.RelationTupleUpdate{

--- a/internal/dispatch/graph/reachableresources_test.go
+++ b/internal/dispatch/graph/reachableresources_test.go
@@ -7,6 +7,7 @@ import (
 	"slices"
 	"sort"
 	"strconv"
+	"sync"
 	"testing"
 
 	"github.com/stretchr/testify/require"
@@ -978,12 +979,13 @@ type breakingDatastore struct {
 
 func (bds breakingDatastore) SnapshotReader(rev datastore.Revision) datastore.Reader {
 	delegate := bds.Datastore.SnapshotReader(rev)
-	return &breakingReader{delegate, 0}
+	return &breakingReader{Reader: delegate, counter: 0, lock: sync.Mutex{}}
 }
 
 type breakingReader struct {
 	datastore.Reader
 	counter int
+	lock    sync.Mutex
 }
 
 func (br *breakingReader) ReverseQueryRelationships(
@@ -991,8 +993,11 @@ func (br *breakingReader) ReverseQueryRelationships(
 	subjectsFilter datastore.SubjectsFilter,
 	options ...options.ReverseQueryOptionsOption,
 ) (datastore.RelationshipIterator, error) {
+	br.lock.Lock()
 	br.counter++
-	if br.counter > 1 {
+	current := br.counter
+	br.lock.Unlock()
+	if current > 1 {
 		return nil, fmt.Errorf("some sort of error")
 	}
 	return br.Reader.ReverseQueryRelationships(ctx, subjectsFilter, options...)
@@ -1392,12 +1397,13 @@ type cancelingDatastore struct {
 
 func (cds cancelingDatastore) SnapshotReader(rev datastore.Revision) datastore.Reader {
 	delegate := cds.Datastore.SnapshotReader(rev)
-	return &cancelingReader{delegate, 0}
+	return &cancelingReader{delegate, 0, sync.Mutex{}}
 }
 
 type cancelingReader struct {
 	datastore.Reader
 	counter int
+	lock    sync.Mutex
 }
 
 func (cr *cancelingReader) ReverseQueryRelationships(
@@ -1405,8 +1411,12 @@ func (cr *cancelingReader) ReverseQueryRelationships(
 	subjectsFilter datastore.SubjectsFilter,
 	options ...options.ReverseQueryOptionsOption,
 ) (datastore.RelationshipIterator, error) {
+	cr.lock.Lock()
 	cr.counter++
-	if cr.counter > 1 {
+	current := cr.counter
+	cr.lock.Unlock()
+
+	if current > 1 {
 		return nil, context.Canceled
 	}
 	return cr.Reader.ReverseQueryRelationships(ctx, subjectsFilter, options...)
@@ -1449,17 +1459,17 @@ func TestReachableResourcesWithCachingInParallelTest(t *testing.T) {
 		require.New(t),
 	)
 
-	dispatcher := NewLocalOnlyDispatcher(50)
-	cachingDispatcher, err := caching.NewCachingDispatcher(caching.DispatchTestCache(t), false, "", &keys.CanonicalKeyHandler{})
-	require.NoError(t, err)
-
-	cachingDispatcher.SetDelegate(dispatcher)
-
 	g := errgroup.Group{}
 	for i := 0; i < 100; i++ {
 		g.Go(func() error {
 			ctx := log.Logger.WithContext(datastoremw.ContextWithHandle(context.Background()))
 			require.NoError(t, datastoremw.SetInContext(ctx, ds))
+
+			dispatcher := NewLocalOnlyDispatcher(50)
+			cachingDispatcher, err := caching.NewCachingDispatcher(caching.DispatchTestCache(t), false, "", &keys.CanonicalKeyHandler{})
+			require.NoError(t, err)
+
+			cachingDispatcher.SetDelegate(dispatcher)
 
 			stream := dispatch.NewCollectingDispatchStream[*v1.DispatchReachableResourcesResponse](ctx)
 			err = cachingDispatcher.DispatchReachableResources(&v1.DispatchReachableResourcesRequest{

--- a/internal/graph/cursors.go
+++ b/internal/graph/cursors.go
@@ -358,9 +358,12 @@ func withInternalParallelizedStreamingIterableInCursor[T any, Q any](
 		taskIndex := taskIndex
 		item := item
 		tr.Add(func(ctx context.Context) error {
+			stream.lock.Lock()
 			if ci.limits.hasExhaustedLimit() {
+				stream.lock.Unlock()
 				return nil
 			}
+			stream.lock.Unlock()
 
 			ici, err := getItemCursor(taskIndex)
 			if err != nil {

--- a/magefiles/test.go
+++ b/magefiles/test.go
@@ -25,7 +25,7 @@ func (t Test) All() error {
 // Unit Runs the unit tests
 func (Test) Unit() error {
 	fmt.Println("running unit tests")
-	return goTest("./...", "-tags", "ci,skipintegrationtests", "-timeout", "10m")
+	return goTest("./...", "-tags", "ci,skipintegrationtests", "-race", "-timeout", "10m")
 }
 
 // Image Run tests that run the built image

